libsnark/zk_proof_systems/plonk/tests/bls12_381_test_vector_transcript_hasher.cpp

```diff
@@ -0,0 +1,138 @@
+        example.nu,
+        example.u,
+    };
+}
+
+void bls12_381_test_vector_transcript_hasher::buffer_clear()
+{
+    this->buffer.clear();
+}
+
+size_t bls12_381_test_vector_transcript_hasher::buffer_size()
+{
+    return this->buffer.size();
+}
+
+void bls12_381_test_vector_transcript_hasher::add_element(
+    const libff::Fr<libff::bls12_381_pp> &element)
+{
+    // convert the Fr element into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::field_write<libff::encoding_binary, libff::form_plain>(
+            element, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+void bls12_381_test_vector_transcript_hasher::add_element(
+    const libff::G1<libff::bls12_381_pp> &element)
+{
+    libff::G1<libff::bls12_381_pp> element_aff(element);
+    element_aff.to_affine_coordinates();
+
+    // convert the affine coordinates of the curve point into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::group_write<
+            libff::encoding_binary,
+            libff::form_plain,
+            libff::compression_off>(element_aff, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+void bls12_381_test_vector_transcript_hasher::add_element(
+    const libff::G2<libff::bls12_381_pp> &element)
+{
+    libff::G2<libff::bls12_381_pp> element_aff(element);
+    element_aff.to_affine_coordinates();
+
+    // convert the affine coordinates of the curve point into a string
+    std::string str;
+    {
+        std::ostringstream ss;
+        libff::group_write<
+            libff::encoding_binary,
+            libff::form_plain,
+            libff::compression_off>(element_aff, ss);
+        str = ss.str();
+    }
+    // copy the string as a sequence of uint8_t elements at the end of
+    // the buffer
+    std::copy(str.begin(), str.end(), std::back_inserter(this->buffer));
+}
+
+libff::Fr<libff::bls12_381_pp> bls12_381_test_vector_transcript_hasher::
+    get_hash()
+{
+    size_t buffer_len = this->buffer.size();
+
+    // find the matching index
+    size_t i = 0;
+    while (buffer_len != length_array[i]) {
+        ++i;
+        if (i >= length_array.size()) {
+            // If we are here, then the hasher buffer has invalid length so
+            // throw an exception
+            throw std::logic_error(
+                "Error: invalid length of transcript hasher buffer");
+        }
+    }
+
+    printf(
+        "[%s:%d] buffer_len %d: %s\n",
+        __FILE__,
+        __LINE__,
+        (int)buffer_len,
+        challenge_array[i].c_str());
+
+    const libff::Fr<libff::bls12_381_pp> challenge = hash_values[i]; // beta
+
+    return challenge;
+}
```

Comment:
> The reason would be to prevent an infinite loop. Or would that be prevented anyway due to the exception?

I would just keep it simple.  Use a for loop to find the right entry, and return.  If you leave the for loop, you didn't find a match.

```suggestion
libff::Fr<libff::bls12_381_pp> bls12_381_test_vector_transcript_hasher::            
    get_hash()                                                                      
{                                                                                   
    const size_t buffer_len = this->buffer.size();                                  
                                                                                    
    // find the matching index and return the corresponding challenge.              
    for (size_t i = 0; i < length_array.size(); ++i) {                              
        if (buffer_len == length_array[i]) {                                        
            printf(                                                                 
                "[%s:%d] buffer_len %d: %s\n",                                      
                __FILE__,                                                           
                __LINE__,                                                           
                (int)buffer_len,                                                    
                challenge_array[i].c_str());                                        
                                                                                    
            return hash_values[i];                                                  
        }                                                                           
    }                                                                               
                                                                                    
    // If we are here, then the hasher buffer has invalid length so                 
    // throw an exception                                                           
    throw std::logic_error("Error: invalid length of transcript hasher buffer");    
}                                                                                   ```

Reply:
Done.